packages/braid-design-system/src/lib/components/MenuRenderer/MenuRenderer.tsx

```diff
@@ -120,85 +146,109 @@ export const MenuRenderer = ({
-          };
-        }
-        default:
+  const [
+    {
+      onClient: hasRenderedOnClient,
+      open,
+      highlightIndex,
+      closeReason,
+      triggerPosition,
+    },
+    dispatch,
+  ] = useReducer((state: State, action: Action): State => {
+    switch (action.type) {
+      case CLIENT_ENVIRONMENT: {
+        return { ...state, onClient: true };
+      }
+      case MENU_TRIGGER_UP:
+      case MENU_ITEM_UP: {
+        return {
+          ...state,
+          open: true,
+          closeReason: CLOSE_REASON_EXIT,
+          highlightIndex: getNextIndex(-1, state.highlightIndex, itemCount),
+          triggerPosition: buttonRef && getPosition(buttonRef.current),
+        };
+      }
+      case MENU_TRIGGER_DOWN:
+      case MENU_ITEM_DOWN: {
+        return {
+          ...state,
+          open: true,
+          closeReason: CLOSE_REASON_EXIT,
+          highlightIndex: getNextIndex(1, state.highlightIndex, itemCount),
+          triggerPosition: buttonRef && getPosition(buttonRef.current),
+        };
+      }
+      case BACKDROP_CLICK:
+      case MENU_TRIGGER_ESCAPE:
+      case MENU_TRIGGER_TAB:
+      case MENU_ITEM_ESCAPE:
+      case MENU_ITEM_TAB: {
+        return {
+          ...state,
+          open: false,
+          closeReason: CLOSE_REASON_EXIT,
+          highlightIndex: CLOSED_INDEX,
+        };
+      }
+      case MENU_ITEM_ENTER:
+      case MENU_ITEM_SPACE:
+      case MENU_ITEM_CLICK: {
+        // Don't close the menu if the user clicked a "form element" item, e.g. checkbox
+        if ('formElement' in action && action.formElement) {
           return state;
+        }
+
+        return {
+          ...state,
+          open: false,
+          closeReason: {
+            reason: 'selection',
+            index: action.index,
+            id: action.id,
+          },
+          highlightIndex: CLOSED_INDEX,
+        };
       }
-    },
-    initialState,
-  );
+      case MENU_ITEM_HOVER: {
+        return { ...state, highlightIndex: action.value };
+      }
+      case MENU_TRIGGER_ENTER:
+      case MENU_TRIGGER_SPACE: {
+        const nextOpen = !state.open;
+        return {
+          ...state,
+          open: nextOpen,
+          closeReason: CLOSE_REASON_EXIT,
+          highlightIndex: nextOpen ? 0 : CLOSED_INDEX,
+          triggerPosition: buttonRef && getPosition(buttonRef.current),
+        };
+      }
+      case MENU_TRIGGER_CLICK: {
+        const nextOpen = !state.open;
+
+        return {
+          ...state,
+          open: nextOpen,
+          closeReason: CLOSE_REASON_EXIT,
+          triggerPosition: buttonRef && getPosition(buttonRef.current),
+        };
+      }
+      case WINDOW_RESIZE: {
+        return {
+          ...state,
+          triggerPosition: buttonRef && getPosition(buttonRef.current),
+        };
+      }
+      default:
+        return state;
+    }
+  }, initialState);
+
+  useEffect(() => {
+    dispatch({ type: CLIENT_ENVIRONMENT });
```

Comment:
I would suggest moving this state from the reducer to a local boolean ref.

Reply:
Have removed this state as part of updating the Menu to only render on `open`